### PR TITLE
Allow setting N=0 to select gridpoints automatically

### PR DESCRIPTION
**Pull request by hungpham2511**
Changes in this PRs:
- Implement an algorithm to choose gridpoint automatically
- Integrate to TOPPRA with N=0

Checklists:
- [ ] Update HISTORY.md with a single line describing this PR

### CODE REVIEW

**Comment by kchenna-nimble** [comment]
I ported this snippet from drake library ([link](https://github.com/RobotLocomotion/drake/blob/master/multibody/optimization/toppra.cc#L17-L60)). It uses linked list instead of vector. It significantly reduces time complexity of function since we don't have to call sort. 

```
toppra::Vector calculateGridPoints(const toppra::GeometricPathPtr path, const double max_err_threshold,
                                   const int max_iteration, const double max_seg_length, const int min_nb_points) {
  auto path_interval = path->pathInterval();
  std::forward_list<double> gridpts{path_interval(0), path_interval(1)};
  int num_grid_points = 2;
  for (int iter = 0; iter < max_iteration; iter++) {
    bool add_points{false};
    for (auto point = gridpts.begin(); std::next(point, 1) != gridpts.end(); point++) {
      const double mid_pt = 0.5 * (*point + *(std::next(point, 1)));
      const double dist = *(std::next(point, 1)) - *point;
      if (dist > max_seg_length) {
        add_points = true;
        gridpts.emplace_after(point, mid_pt);
        num_grid_points++;
        point++;
        continue;
      }

      const double error = 0.5 * std::pow(dist, 2) * path->eval_single(mid_pt, 2).cwiseAbs().maxCoeff();
      if (error > max_err_threshold) {
        add_points = true;
        gridpts.emplace_after(point, mid_pt);
        num_grid_points++;
        point++;
        continue;
      }
    }
    if (!add_points) {
      break;
    }
  }
  while (num_grid_points < min_nb_points) {
    for (auto point = gridpts.begin(); std::next(point, 1) != gridpts.end(); std::advance(point, 2)) {
      const double mid_pt = 0.5 * (*point + *(std::next(point, 1)));
      gridpts.emplace_after(point, mid_pt);
      num_grid_points++;
    }
  }
  std::vector<double> result(gridpts.begin(), gridpts.end());
  return Eigen::Map<toppra::Vector, Eigen::Unaligned>(result.data(), result.size());
}
```

**Comment by hashb** [comment]
I ported this snippet from drake library ([link](https://github.com/RobotLocomotion/drake/blob/master/multibody/optimization/toppra.cc#L17-L60)). It uses linked list instead of vector. It significantly reduces time complexity of function since we don't have to call sort. Would be great to have it integrated into toppra library.

```
toppra::Vector calculateGridPoints(const toppra::GeometricPathPtr path, const double max_err_threshold,
                                   const int max_iteration, const double max_seg_length, const int min_nb_points) {
  auto path_interval = path->pathInterval();
  std::forward_list<double> gridpts{path_interval(0), path_interval(1)};
  int num_grid_points = 2;
  for (int iter = 0; iter < max_iteration; iter++) {
    bool add_points{false};
    for (auto point = gridpts.begin(); std::next(point, 1) != gridpts.end(); point++) {
      const double mid_pt = 0.5 * (*point + *(std::next(point, 1)));
      const double dist = *(std::next(point, 1)) - *point;
      if (dist > max_seg_length) {
        add_points = true;
        gridpts.emplace_after(point, mid_pt);
        num_grid_points++;
        point++;
        continue;
      }

      const double error = 0.5 * std::pow(dist, 2) * path->eval_single(mid_pt, 2).cwiseAbs().maxCoeff();
      if (error > max_err_threshold) {
        add_points = true;
        gridpts.emplace_after(point, mid_pt);
        num_grid_points++;
        point++;
        continue;
      }
    }
    if (!add_points) {
      break;
    }
  }
  while (num_grid_points < min_nb_points) {
    for (auto point = gridpts.begin(); std::next(point, 1) != gridpts.end(); std::advance(point, 2)) {
      const double mid_pt = 0.5 * (*point + *(std::next(point, 1)));
      gridpts.emplace_after(point, mid_pt);
      num_grid_points++;
    }
  }
  std::vector<double> result(gridpts.begin(), gridpts.end());
  return Eigen::Map<toppra::Vector, Eigen::Unaligned>(result.data(), result.size());
}
```

**Comment by hungpham2511** [comment]
Nice work! Thanks.

**Comment by hungpham2511** [comment]
@jmirabel let me know if you would like to have a look. Otherwise I will merge this in a few days.

File: cpp/src/toppra/algorithm/toppra.cpp [hunk 1]
```diff
@@ -21,7 +21,8 @@ ReturnCode TOPPRA::computeForwardPass(value_type vel_start) {
   auto deltas = m_solver->deltas();
   Bound x, x_next;
   m_data.parametrization(0) = vel_start;
-  for (std::size_t i = 0; i < m_N; i++) {
+  for (std::size_t i = 0; i < m_data.parametrization.size() - 1; i++) {
```
Comment (jmirabel) [inline cpp/src/toppra/algorithm/toppra.cpp:24]: Is there a specific reason for this change ?

File: cpp/src/toppra/geometric_path.cpp [hunk 2]
```diff
@@ -10,4 +13,50 @@ Vectors GeometricPath::eval(const Vector &positions, int order) const {
...
     return outputs;
   };
+
+
+Vector GeometricPath::proposeGridpoints(double max_err_threshold, int max_iteration, double max_seg_length, int min_nb_points) const {
+  std::forward_list<value_type> gridpoints {pathInterval()[0], pathInterval()[1]};
+  int nb_gridpoints = 2;
+
+  // Add points according to error threshold
+  for (auto iter=0; iter < max_iteration; iter++){
```
Comment (jmirabel) [inline cpp/src/toppra/geometric_path.cpp]: Hum, rather strange use of `auto`...
```suggestion
  for (int iter=0; iter < max_iteration; iter++){
```

File: cpp/src/toppra/geometric_path.cpp [hunk 3]
```diff
@@ -10,4 +13,50 @@ Vectors GeometricPath::eval(const Vector &positions, int order) const {
...
+      auto dist = (*std::next(point, 1) - *point);
+
+      if (dist > max_seg_length){
+        gridpoints.emplace_after(point, p_mid);
+        add_new_points = true;
+        nb_gridpoints ++;
+        continue;
+      }
+
+      auto max_err = (0.5 * eval_single(p_mid, 2) * dist * dist).cwiseAbs().maxCoeff();
```
Comment (jmirabel) [inline cpp/src/toppra/geometric_path.cpp:38]: Maybe you can comment what this line does. I don't understand it.

**Review by jmirabel — COMMENTED** [review]
Do you expect some kind of regularity in the spacing between points ? I don't completely grasp the logic.

**Comment by hungpham2511** [comment]
> Do you expect some kind of regularity in the spacing between points ? I don't completely grasp the logic.

It's a port of the propose gridpoint procedure in the  Python code. I will copy the docstring over. That should explain what is happening.

File: cpp/src/toppra/geometric_path.cpp (hunk 3 above)
Comment (hungpham2511) [reply cpp/src/toppra/geometric_path.cpp:38]: It's the maximum interpolation error.

File: cpp/src/toppra/algorithm/toppra.cpp (hunk 1 above)
Comment (hungpham2511) [reply cpp/src/toppra/algorithm/toppra.cpp:24]: Strictly speaking, checking the stopping condition based on the parametrization vector is more accurate.

This change is required as now we allow user to pass in `N=0` to use the automatically generated grid-points.

File: cpp/src/toppra/geometric_path.cpp (hunk 3 above)
Comment (hungpham2511) [reply cpp/src/toppra/geometric_path.cpp:38]: Hi @jmirabel I have updated the docstring. Hope it's clearer now.

**Review by jmirabel — APPROVED** [review]
Thanks.

**Pull request merged**